File: typescript/tests/storage-memory/template-cycle.test.ts
/**
 * Cycle detection on template re-parenting (memory adapter).
 *
 * The storage spec requires that any update placing a template's parent
 * chain into a cycle is rejected at write time with `TemplateCycle`. This
 * test exercises the memory adapter; sqlite + postgres tests live alongside
 * their adapter directories.
 */

import { describe, expect, it } from 'bun:test';

import { MemoryStorage } from '../../src/storage/memory/index.ts';
import type { LicenseTemplateInput } from '../../src/storage/types.ts';
import type { UUIDv7 } from '../../src/types.ts';

function freshTemplateInput(name: string, parent_id: UUIDv7 | null = null): LicenseTemplateInput {
  return {
    scope_id: null,
    parent_id,
    name,
    max_usages: 5,
    trial_duration_sec: 0,
    trial_cooldown_sec: null,
    grace_duration_sec: 0,
    force_online_after_sec: null,
    entitlements: {},
    meta: {},
  };
}

describe('memory adapter — template cycle detection', () => {
  it('rejects direct self-cycle on update (parent_id = self.id)', async () => {
    const s = new MemoryStorage();
    const a = await s.createTemplate(freshTemplateInput('a'));
    await expect(s.updateTemplate(a.id, { parent_id: a.id })).rejects.toMatchObject({
      code: 'TemplateCycle',
    });
    // Row unchanged.
    const after = await s.getTemplate(a.id);
    expect(after?.parent_id).toBeNull();
  });

  it('rejects indirect cycle: a → b → c, then update a.parent_id to c', async () => {
    const s = new MemoryStorage();
    const a = await s.createTemplate(freshTemplateInput('a'));
    const b = await s.createTemplate(freshTemplateInput('b', a.id));
    const c = await s.createTemplate(freshTemplateInput('c', b.id));
    await expect(s.updateTemplate(a.id, { parent_id: c.id })).rejects.toMatchObject({
      code: 'TemplateCycle',
    });
    // Existing chain intact.
    const aAfter = await s.getTemplate(a.id);
    const bAfter = await s.getTemplate(b.id);
    const cAfter = await s.getTemplate(c.id);
    expect(aAfter?.parent_id).toBeNull();
    expect(bAfter?.parent_id).toBe(a.id);
    expect(cAfter?.parent_id).toBe(b.id);
  });

  it('allows valid re-parenting that does not introduce a cycle', async () => {
    const s = new MemoryStorage();
    const root1 = await s.createTemplate(freshTemplateInput('root1'));
    const root2 = await s.createTemplate(freshTemplateInput('root2'));
    const child = await s.createTemplate(freshTemplateInput('child', root1.id));
    const moved = await s.updateTemplate(child.id, { parent_id: root2.id });
    expect(moved.parent_id).toBe(root2.id);
  });

  it('allows detaching a child by setting parent_id to null', async () => {
    const s = new MemoryStorage();
    const a = await s.createTemplate(freshTemplateInput('a'));
    const b = await s.createTemplate(freshTemplateInput('b', a.id));
    const detached = await s.updateTemplate(b.id, { parent_id: null });
    expect(detached.parent_id).toBeNull();
  });

  it('rejects createTemplate with a parent_id that does not exist', async () => {
    const s = new MemoryStorage();
    const fakeParent = '01939e6f-0000-7000-8000-000000000099' as UUIDv7;
    await expect(s.createTemplate(freshTemplateInput('a', fakeParent))).rejects.toMatchObject({
      code: 'UniqueConstraintViolation',
    });
  });

  it('listTemplates filters by parent_id (including parent_id: null for roots)', async () => {
    const s = new MemoryStorage();
    const root = await s.createTemplate(freshTemplateInput('root'));
    await s.createTemplate(freshTemplateInput('child1', root.id));
    await s.createTemplate(freshTemplateInput('child2', root.id));

    const roots = await s.listTemplates({ parent_id: null }, { limit: 10 });
    expect(roots.items.length).toBe(1);
    expect(roots.items[0]?.name).toBe('root');

    const children = await s.listTemplates({ parent_id: root.id }, { limit: 10 });
    expect(children.items.length).toBe(2);
    expect(children.items.map((t) => t.name).sort()).toEqual(['child1', 'child2']);
  });
});

File: licensing/storage/memory/template_cycle_test.go
package memory_test

import (
	"errors"
	"testing"

	lic "github.com/AnoRebel/licensing/licensing"
	"github.com/AnoRebel/licensing/licensing/storage/memory"
)

// Mirrors typescript/tests/storage-memory/template-cycle.test.ts.

func tmplInput(name string, parentID *string) lic.LicenseTemplateInput {
	return lic.LicenseTemplateInput{
		ScopeID:             nil,
		ParentID:            parentID,
		ForceOnlineAfterSec: nil,
		TrialCooldownSec:    nil,
		Entitlements:        map[string]any{},
		Meta:                map[string]any{},
		Name:                name,
		MaxUsages:           5,
		TrialDurationSec:    0,
		GraceDurationSec:    0,
	}
}

func TestMemory_TemplateCycle_Direct(t *testing.T) {
	s := memory.New(memory.Options{})
	a, err := s.CreateTemplate(tmplInput("a", nil))
	if err != nil {
		t.Fatal(err)
	}
	_, err = s.UpdateTemplate(a.ID, lic.LicenseTemplatePatch{
		ParentID: lic.OptString{Set: true, Value: &a.ID},
	})
	if err == nil {
		t.Fatal("expected TemplateCycle error")
	}
	var lerr *lic.Error
	if !errors.As(err, &lerr) || lerr.Code != lic.CodeTemplateCycle {
		t.Fatalf("expected CodeTemplateCycle, got %v", err)
	}
}

func TestMemory_TemplateCycle_Indirect(t *testing.T) {
	s := memory.New(memory.Options{})
	a, _ := s.CreateTemplate(tmplInput("a", nil))
	b, _ := s.CreateTemplate(tmplInput("b", &a.ID))
	c, _ := s.CreateTemplate(tmplInput("c", &b.ID))
	_, err := s.UpdateTemplate(a.ID, lic.LicenseTemplatePatch{
		ParentID: lic.OptString{Set: true, Value: &c.ID},
	})
	if err == nil {
		t.Fatal("expected TemplateCycle error")
	}
	var lerr *lic.Error
	if !errors.As(err, &lerr) || lerr.Code != lic.CodeTemplateCycle {
		t.Fatalf("expected CodeTemplateCycle, got %v", err)
	}
}

func TestMemory_TemplateCycle_AllowsValidReParent(t *testing.T) {
	s := memory.New(memory.Options{})
	r1, _ := s.CreateTemplate(tmplInput("r1", nil))
	r2, _ := s.CreateTemplate(tmplInput("r2", nil))
	child, _ := s.CreateTemplate(tmplInput("child", &r1.ID))
	moved, err := s.UpdateTemplate(child.ID, lic.LicenseTemplatePatch{
		ParentID: lic.OptString{Set: true, Value: &r2.ID},
	})
	if err != nil {
		t.Fatal(err)
	}
	if moved.ParentID == nil || *moved.ParentID != r2.ID {
		t.Fatalf("re-parent failed: got %v want %s", moved.ParentID, r2.ID)
	}
	// Detach to nil.
	detached, err := s.UpdateTemplate(child.ID, lic.LicenseTemplatePatch{
		ParentID: lic.OptString{Set: true, Value: nil},
	})
	if err != nil {
		t.Fatal(err)
	}
	if detached.ParentID != nil {
		t.Fatalf("expected nil parent_id, got %v", detached.ParentID)
	}
}

func TestMemory_TemplateFilter_ParentID(t *testing.T) {
	s := memory.New(memory.Options{})
	root, _ := s.CreateTemplate(tmplInput("root", nil))
	_, _ = s.CreateTemplate(tmplInput("c1", &root.ID))
	_, _ = s.CreateTemplate(tmplInput("c2", &root.ID))

	roots, err := s.ListTemplates(
		lic.LicenseTemplateFilter{ParentIDSet: true, ParentID: nil},
		lic.PageRequest{Limit: 10},
	)
	if err != nil {
		t.Fatal(err)
	}
	if len(roots.Items) != 1 {
		t.Fatalf("expected 1 root, got %d", len(roots.Items))
	}
	children, err := s.ListTemplates(
		lic.LicenseTemplateFilter{ParentIDSet: true, ParentID: &root.ID},
		lic.PageRequest{Limit: 10},
	)
	if err != nil {
		t.Fatal(err)
	}
	if len(children.Items) != 2 {
		t.Fatalf("expected 2 children, got %d", len(children.Items))
	}
}
